cbmc/proofs/mqtt_state.c

```diff
@@ -545,43 +568,41 @@ void free_IotMqttSubscriptionList( IotListDouble_t *pSub )
  * IotMqttPublishInfo
  ****************************************************************/
 
+void bound_IotMqttPublishInfo( IotMqttPublishInfo_t *pInfo )
+{
+    #ifdef TOPIC_NAME_LENGTH_MAX
+        __CPROVER_assume( pInfo->topicNameLength < TOPIC_NAME_LENGTH_MAX &&
+                          VALID_CBMC_SIZE( pInfo->topicNameLength ) );
+    #else
+      __CPROVER_assume( pInfo->topicNameLength <= UINT16_MAX );
+    #endif
+
+    #ifdef PAYLOAD_LENGTH_MAX
+        __CPROVER_assume( pInfo->payloadLength < PAYLOAD_LENGTH_MAX &&
+                          VALID_CBMC_SIZE( pInfo->payloadLength ) );
+    #else
+      __CPROVER_assume( pInfo->payloadLength <= UINT16_MAX );
+    #endif
+}
+
 IotMqttPublishInfo_t *allocate_IotMqttPublishInfo( IotMqttPublishInfo_t *pInfo )
 {
   if ( pInfo == NULL ) pInfo = malloc_can_fail( sizeof( *pInfo ) );
   if ( pInfo == NULL ) return NULL;
 
+  bound_IotMqttPublishInfo( pInfo );
+
   pInfo->pTopicName = malloc_can_fail( pInfo->topicNameLength );
-  // assumption here is checked below in valid_
-  __CPROVER_assume( VALID_CBMC_SIZE( pInfo->payloadLength ) );
   pInfo->pPayload = malloc_can_fail( pInfo->payloadLength );
   return pInfo;
 }
 
 bool valid_IotMqttPublishInfo( const IotMqttPublishInfo_t *pInfo )
 {
   return
-    pInfo &&
-
+    pInfo != NULL &&
     VALID_QOS(pInfo->qos) &&
-
-    VALID_STRING( pInfo->pTopicName, pInfo->topicNameLength ) &&
-    VALID_CBMC_SIZE( pInfo->topicNameLength ) &&
-    pInfo->pTopicName != NULL &&
-
-    VALID_STRING( pInfo->pPayload, pInfo->payloadLength ) &&
-    VALID_CBMC_SIZE( pInfo->payloadLength ) &&
-#ifdef PAYLOAD_LENGTH_MAX
-    // Some proofs iterate over it
-    pInfo->payloadLength < PAYLOAD_LENGTH_MAX &&
-#endif
-
-    pInfo->retryMs <= IOT_MQTT_RETRY_MS_CEILING &&
-    pInfo->retryMs > 0  &&
-
-    // TODO: experiment with removing these assumptions
-    // topicNameLength is a unint16
-    pInfo->topicNameLength <= UINT16_MAX &&
```

Comment:
We need to be absolutely sure this continues to work, see the comment, this could affect everything from errors to coverage.  Let's not make changes to the mqtt state model gratuitously at this point.